File: desktop/lib/models/vendor.dart
import 'entities.dart';

class VendorContact {
  const VendorContact({
    required this.id,
    required this.name,
    required this.department,
    required this.designation,
    required this.phone,
    required this.mobile,
    required this.email,
    required this.isPrimary,
    required this.status,
  });

  final String id;
  final String name;
  final String department;
  final String designation;
  final String phone;
  final String mobile;
  final String email;
  final bool isPrimary;
  final String status;

  factory VendorContact.fromJson(Json json) => VendorContact(
        id: stringValue(json['id']),
        name: stringValue(json['name']),
        department: stringValue(json['department']),
        designation: stringValue(json['designation']),
        phone: stringValue(json['phone']),
        mobile: stringValue(json['mobile']),
        email: stringValue(json['email']),
        isPrimary: boolValue(json['is_primary']),
        status: stringValue(json['status']),
      );

  Json toJson() => {
        if (id.isNotEmpty) 'id': id,
        'name': name,
        'department': department,
        'designation': designation,
        'phone': phone,
        'mobile': mobile,
        'email': email,
        'is_primary': isPrimary,
        'status': status,
      };
}

class VendorAddress {
  const VendorAddress({
    required this.id,
    required this.addressType,
    required this.addressLine1,
    required this.addressLine2,
    required this.countryId,
    required this.stateId,
    required this.districtId,
    required this.cityId,
    required this.postalCodeId,
    required this.localityId,
    required this.isPrimary,
  });

  final String id;
  final String addressType;
  final String addressLine1;
  final String addressLine2;
  final String countryId;
  final String stateId;
  final String districtId;
  final String cityId;
  final String postalCodeId;
  final String localityId;
  final bool isPrimary;

  factory VendorAddress.fromJson(Json json) => VendorAddress(
        id: stringValue(json['id']),
        addressType: stringValue(json['address_type']),
        addressLine1: stringValue(json['address_line1']),
        addressLine2: stringValue(json['address_line2']),
        countryId: stringValue(json['country_id']),
        stateId: stringValue(json['state_id']),
        districtId: stringValue(json['district_id']),
        cityId: stringValue(json['city_id']),
        postalCodeId: stringValue(json['postal_code_id']),
        localityId: stringValue(json['locality_id']),
        isPrimary: boolValue(json['is_primary']),
      );

  Json toJson() => {
        if (id.isNotEmpty) 'id': id,
        'address_type': addressType,
        'address_line1': addressLine1,
        'address_line2': addressLine2,
        if (countryId.isNotEmpty) 'country_id': countryId,
        if (stateId.isNotEmpty) 'state_id': stateId,
        if (districtId.isNotEmpty) 'district_id': districtId,
        if (cityId.isNotEmpty) 'city_id': cityId,
        if (postalCodeId.isNotEmpty) 'postal_code_id': postalCodeId,
        if (localityId.isNotEmpty) 'locality_id': localityId,
        'is_primary': isPrimary,
      };
}

/// One bank account money is paid into.
class VendorBankAccount {
  const VendorBankAccount({
    required this.id,
    required this.bankName,
    required this.accountName,
    required this.accountNumber,
    required this.ifsc,
    required this.branch,
    required this.upiId,
    required this.swiftCode,
    required this.isPrimary,
  });

  final String id;
  final String bankName;
  final String accountName;
  final String accountNumber;
  final String ifsc;
  final String branch;
  final String upiId;
  final String swiftCode;
  final bool isPrimary;

  factory VendorBankAccount.fromJson(Json json) => VendorBankAccount(
        id: stringValue(json['id']),
        bankName: stringValue(json['bank_name']),
        accountName: stringValue(json['account_name']),
        accountNumber: stringValue(json['account_number']),
        ifsc: stringValue(json['ifsc']),
        branch: stringValue(json['branch']),
        upiId: stringValue(json['upi_id']),
        swiftCode: stringValue(json['swift_code']),
        isPrimary: boolValue(json['is_primary']),
      );

  Json toJson() => {
        if (id.isNotEmpty) 'id': id,
        'bank_name': bankName,
        'account_name': accountName,
        'account_number': accountNumber,
        'ifsc': ifsc.isEmpty ? null : ifsc,
        'branch': branch.isEmpty ? null : branch,
        'upi_id': upiId.isEmpty ? null : upiId,
        'swift_code': swiftCode.isEmpty ? null : swiftCode,
        'is_primary': isPrimary,
      };
}

/// One set of a vendor's registrations.
///
/// The vendor record carries a `gstin` and a `pan` of its own; these rows are
/// the per-registration detail — a vendor trading from two states has two
/// GSTINs, and only one of them is primary.
class VendorTaxDetail {
  const VendorTaxDetail({
    required this.id,
    required this.gstin,
    required this.pan,
    required this.tan,
    required this.fssai,
    required this.drugLicense,
    required this.importExportCode,
    required this.isPrimary,
  });

  final String id;
  final String gstin;
  final String pan;
  final String tan;
  final String fssai;
  final String drugLicense;
  final String importExportCode;
  final bool isPrimary;

  factory VendorTaxDetail.fromJson(Json json) => VendorTaxDetail(
        id: stringValue(json['id']),
        gstin: stringValue(json['gstin']),
        pan: stringValue(json['pan']),
        tan: stringValue(json['tan']),
        fssai: stringValue(json['fssai']),
        drugLicense: stringValue(json['drug_license']),
        importExportCode: stringValue(json['import_export_code']),
        isPrimary: boolValue(json['is_primary']),
      );

  Json toJson() => {
        if (id.isNotEmpty) 'id': id,
        'gstin': gstin.isEmpty ? null : gstin,
        'pan': pan.isEmpty ? null : pan,
        'tan': tan.isEmpty ? null : tan,
        'fssai': fssai.isEmpty ? null : fssai,
        'drug_license': drugLicense.isEmpty ? null : drugLicense,
        'import_export_code':
            importExportCode.isEmpty ? null : importExportCode,
        'is_primary': isPrimary,
      };
}

/// One note kept against a vendor.
class VendorNote {
  const VendorNote({
    required this.id,
    required this.note,
    required this.noteType,
  });

  final String id;
  final String note;
  final String noteType;

  factory VendorNote.fromJson(Json json) => VendorNote(
        id: stringValue(json['id']),
        note: stringValue(json['note']),
        noteType: stringValue(json['note_type']),
      );

  Json toJson() => {
        if (id.isNotEmpty) 'id': id,
        'note': note,
        'note_type': noteType.isEmpty ? 'GENERAL' : noteType,
      };
}

class Vendor {
  const Vendor({
    required this.id,
    this.version = 0,
    required this.firmId,
    required this.code,
    required this.name,
    required this.legalName,
    required this.displayName,
    required this.categoryId,
    required this.typeId,
    required this.status,
    required this.businessProfileId,
    required this.gstRegistration,
    required this.gstin,
    required this.pan,
    required this.licenseNumber,
    required this.registrationNumber,
    required this.website,
    required this.email,
    required this.phone,
    required this.mobile,
    required this.remarks,
    required this.businessAttributes,
    required this.createdAt,
    required this.updatedAt,
    required this.isDeleted,
    required this.contacts,
    required this.addresses,
    this.bankAccounts = const [],
    this.taxDetails = const [],
    this.notes = const [],
  });

  final String id;

  /// The optimistic-concurrency version this record was read at, sent back
  /// as `If-Match` on save so a concurrent edit is refused rather than
  /// silently overwritten. Zero means the server published none, and the
  /// save then carries no precondition.
  final int version;
  final String firmId;
  final String code;
  final String name;
  final String legalName;
  final String displayName;
  final String categoryId;
  final String typeId;
  final String status;
  final String businessProfileId;
  final bool gstRegistration;
  final String gstin;
  final String pan;
  final String licenseNumber;
  final String registrationNumber;
  final String website;
  final String email;
  final String phone;
  final String mobile;
  final String remarks;
  final Json businessAttributes;
  final String createdAt;
  final String updatedAt;
  final bool isDeleted;
  final List<VendorContact> contacts;
  final List<VendorAddress> addresses;
  final List<VendorBankAccount> bankAccounts;
  final List<VendorTaxDetail> taxDetails;
  final List<VendorNote> notes;

  factory Vendor.fromJson(Json json) => Vendor(
        id: stringValue(json['id']),
        version: (json['version'] as num?)?.toInt() ?? 0,
        firmId: stringValue(json['firm_id']),
        code: stringValue(json['code']),
        name: stringValue(json['name']),
        legalName: stringValue(json['legal_name']),
        displayName: stringValue(json['display_name']),
        categoryId: stringValue(json['category_id']),
        typeId: stringValue(json['type_id']),
        status: stringValue(json['status']),
        businessProfileId: stringValue(json['business_profile_id']),
        gstRegistration: boolValue(json['gst_registration']),
        gstin: stringValue(json['gstin']),
        pan: stringValue(json['pan']),
        licenseNumber: stringValue(json['license_number']),
        registrationNumber: stringValue(json['registration_number']),
        website: stringValue(json['website']),
        email: stringValue(json['email']),
        phone: stringValue(json['phone']),
        mobile: stringValue(json['mobile']),
        remarks: stringValue(json['remarks']),
        businessAttributes: json['business_attributes'] is Map
            ? Map<String, dynamic>.from(json['business_attributes'] as Map)
            : const {},
        createdAt: stringValue(json['created_at']),
        updatedAt: stringValue(json['updated_at']),
        isDeleted: boolValue(json['is_deleted']),
        contacts:
            _objects(json['contacts']).map(VendorContact.fromJson).toList(),
        addresses:
            _objects(json['addresses']).map(VendorAddress.fromJson).toList(),
        bankAccounts: _objects(json['bank_accounts'])
            .map(VendorBankAccount.fromJson)
            .toList(),
        taxDetails:
            _objects(json['tax_details']).map(VendorTaxDetail.fromJson).toList(),
        notes: _objects(json['notes']).map(VendorNote.fromJson).toList(),
      );
}

class VendorQuery {
  const VendorQuery({
    this.status,
    this.categoryId,
    this.typeId,
    this.businessProfileId,
    this.includeDeleted = false,
  });

  final String? status;
  final String? categoryId;
  final String? typeId;
  final String? businessProfileId;
  final bool includeDeleted;

  Map<String, String> toQuery() => {
        if (status?.isNotEmpty == true) 'status': status!,
        if (categoryId?.isNotEmpty == true) 'category_id': categoryId!,
        if (typeId?.isNotEmpty == true) 'type_id': typeId!,
        if (businessProfileId?.isNotEmpty == true)
          'business_profile_id': businessProfileId!,
        if (includeDeleted) 'include_deleted': 'true',
      };
}

List<Json> _objects(dynamic value) => value is List
    ? value
        .whereType<Map>()
        .map((item) => Map<String, dynamic>.from(item))
        .toList()
    : const [];

File: backend/app/vendors/models/vendor.py
"""Firm-scoped vendor and child persistence models."""

from typing import ClassVar
from uuid import UUID

from sqlalchemy import (
    JSON,
    Boolean,
    ForeignKey,
    Index,
    String,
    Text,
    UniqueConstraint,
    and_,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.business.models import AttributeEntityType, AttributeValueBase
from app.core.database.entity import BaseEntity
from app.core.database.types import UUIDType


class VendorCategory(BaseEntity):
    """Persist a reusable vendor category per firm."""

    __tablename__ = "vendor_categories"
    __table_args__ = (
        UniqueConstraint("firm_id", "code", name="UQ_vendor_categories_firm_code"),
        UniqueConstraint("firm_id", "name", name="UQ_vendor_categories_firm_name"),
    )

    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    code: Mapped[str] = mapped_column(String(50), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    is_active: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default="true"
    )


class VendorType(BaseEntity):
    """Persist a reusable vendor type per firm."""

    __tablename__ = "vendor_types"
    __table_args__ = (
        UniqueConstraint("firm_id", "code", name="UQ_vendor_types_firm_code"),
        UniqueConstraint("firm_id", "name", name="UQ_vendor_types_firm_name"),
    )

    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    code: Mapped[str] = mapped_column(String(50), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    is_active: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default="true"
    )


class Vendor(BaseEntity):
    """Represent one vendor master owned by a firm."""

    __tablename__ = "vendors"
    __table_args__ = (
        UniqueConstraint("firm_id", "code", name="UQ_vendors_firm_code"),
        UniqueConstraint("firm_id", "gstin", name="UQ_vendors_firm_gstin"),
        Index("IX_vendors_firm_name", "firm_id", "name"),
        Index("IX_vendors_firm_status", "firm_id", "status"),
    )

    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    code: Mapped[str] = mapped_column(String(50), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    legal_name: Mapped[str | None] = mapped_column(String(200))
    display_name: Mapped[str] = mapped_column(String(200), nullable=False)
    category_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("vendor_categories.id", ondelete="RESTRICT")
    )
    type_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("vendor_types.id", ondelete="RESTRICT")
    )
    business_profile_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("business_profiles.id", ondelete="RESTRICT")
    )
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="ACTIVE", server_default="ACTIVE"
    )
    gst_registration: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    gstin: Mapped[str | None] = mapped_column(String(32))
    pan: Mapped[str | None] = mapped_column(String(32))
    license_number: Mapped[str | None] = mapped_column(String(64))
    registration_number: Mapped[str | None] = mapped_column(String(64))
    website: Mapped[str | None] = mapped_column(String(500))
    email: Mapped[str | None] = mapped_column(String(320))
    phone: Mapped[str | None] = mapped_column(String(20))
    mobile: Mapped[str | None] = mapped_column(String(20))
    remarks: Mapped[str | None] = mapped_column(Text)
    business_attributes: Mapped[dict[str, object]] = mapped_column(
        JSON, nullable=False, default=dict
    )

    contacts: Mapped[list["VendorContact"]] = relationship(
        back_populates="vendor",
        cascade="save-update, merge",
        primaryjoin=lambda: and_(
            Vendor.id == VendorContact.vendor_id,
            VendorContact.is_deleted.is_(False),
        ),
        lazy="selectin",
        order_by="VendorContact.created_at",
    )
    addresses: Mapped[list["VendorAddress"]] = relationship(
        back_populates="vendor",
        cascade="save-update, merge",
        primaryjoin=lambda: and_(
            Vendor.id == VendorAddress.vendor_id,
            VendorAddress.is_deleted.is_(False),
        ),
        lazy="selectin",
        order_by="VendorAddress.created_at",
    )
    bank_accounts: Mapped[list["VendorBankAccount"]] = relationship(
        back_populates="vendor",
        cascade="save-update, merge",
        primaryjoin=lambda: and_(
            Vendor.id == VendorBankAccount.vendor_id,
            VendorBankAccount.is_deleted.is_(False),
        ),
        lazy="selectin",
        order_by="VendorBankAccount.created_at",
    )
    tax_details: Mapped[list["VendorTaxDetail"]] = relationship(
        back_populates="vendor",
        cascade="save-update, merge",
        primaryjoin=lambda: and_(
            Vendor.id == VendorTaxDetail.vendor_id,
            VendorTaxDetail.is_deleted.is_(False),
        ),
        lazy="selectin",
        order_by="VendorTaxDetail.created_at",
    )
    notes: Mapped[list["VendorNote"]] = relationship(
        back_populates="vendor",
        cascade="save-update, merge",
        primaryjoin=lambda: and_(
            Vendor.id == VendorNote.vendor_id,
            VendorNote.is_deleted.is_(False),
        ),
        lazy="selectin",
        order_by="VendorNote.created_at",
    )
    attachments: Mapped[list["VendorAttachment"]] = relationship(
        back_populates="vendor",
        cascade="save-update, merge",
        primaryjoin=lambda: and_(
            Vendor.id == VendorAttachment.vendor_id,
            VendorAttachment.is_deleted.is_(False),
        ),
        lazy="selectin",
        order_by="VendorAttachment.created_at",
    )


class VendorContact(BaseEntity):
    """Represent one vendor contact person."""

    __tablename__ = "vendor_contacts"

    vendor_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("vendors.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    department: Mapped[str | None] = mapped_column(String(100))
    designation: Mapped[str | None] = mapped_column(String(100))
    phone: Mapped[str | None] = mapped_column(String(20))
    mobile: Mapped[str | None] = mapped_column(String(20))
    email: Mapped[str | None] = mapped_column(String(320))
    is_primary: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="ACTIVE", server_default="ACTIVE"
    )

    vendor: Mapped[Vendor] = relationship(back_populates="contacts")


class VendorAddress(BaseEntity):
    """Represent one vendor address referencing geo masters."""

    __tablename__ = "vendor_addresses"
    __table_args__ = (Index("IX_vendor_addresses_vendor_city", "vendor_id", "city_id"),)

    vendor_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("vendors.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    address_type: Mapped[str] = mapped_column(String(20), nullable=False)
    address_line1: Mapped[str] = mapped_column(String(250), nullable=False)
    address_line2: Mapped[str | None] = mapped_column(String(250))
    country_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_countries.id", ondelete="RESTRICT")
    )
    state_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_states.id", ondelete="RESTRICT")
    )
    district_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_districts.id", ondelete="RESTRICT")
    )
    city_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_cities.id", ondelete="RESTRICT")
    )
    postal_code_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_postal_codes.id", ondelete="RESTRICT")
    )
    locality_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("geo_localities.id", ondelete="RESTRICT")
    )
    is_primary: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )

    vendor: Mapped[Vendor] = relationship(back_populates="addresses")


class VendorBankAccount(BaseEntity):
    """Represent one vendor bank account."""

    __tablename__ = "vendor_bank_accounts"
    __table_args__ = (
        Index("IX_vendor_bank_accounts_vendor_primary", "vendor_id", "is_primary"),
    )

    vendor_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("vendors.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    bank_name: Mapped[str] = mapped_column(String(150), nullable=False)
    account_name: Mapped[str] = mapped_column(String(150), nullable=False)
    account_number: Mapped[str] = mapped_column(String(64), nullable=False)
    ifsc: Mapped[str | None] = mapped_column(String(16))
    branch: Mapped[str | None] = mapped_column(String(120))
    upi_id: Mapped[str | None] = mapped_column(String(120))
    swift_code: Mapped[str | None] = mapped_column(String(16))
    is_primary: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )

    vendor: Mapped[Vendor] = relationship(back_populates="bank_accounts")


class VendorTaxDetail(BaseEntity):
    """Represent one vendor tax detail set."""

    __tablename__ = "vendor_tax_details"
    __table_args__ = (
        Index("IX_vendor_tax_details_vendor_primary", "vendor_id", "is_primary"),
    )

    vendor_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("vendors.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    gstin: Mapped[str | None] = mapped_column(String(32))
    pan: Mapped[str | None] = mapped_column(String(32))
    tan: Mapped[str | None] = mapped_column(String(32))
    fssai: Mapped[str | None] = mapped_column(String(32))
    drug_license: Mapped[str | None] = mapped_column(String(64))
    import_export_code: Mapped[str | None] = mapped_column(String(32))
    extra_fields: Mapped[dict[str, object]] = mapped_column(
        JSON, nullable=False, default=dict
    )
    is_primary: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default="false"
    )

    vendor: Mapped[Vendor] = relationship(back_populates="tax_details")


class VendorAttachment(BaseEntity):
    """Represent one vendor attachment metadata row."""

    __tablename__ = "vendor_attachments"

    vendor_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("vendors.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    file_name: Mapped[str] = mapped_column(String(255), nullable=False)
    file_url: Mapped[str] = mapped_column(String(1000), nullable=False)
    mime_type: Mapped[str | None] = mapped_column(String(120))
    description: Mapped[str | None] = mapped_column(Text)

    vendor: Mapped[Vendor] = relationship(back_populates="attachments")


class VendorNote(BaseEntity):
    """Represent one vendor note/history item."""

    __tablename__ = "vendor_notes"

    vendor_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("vendors.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    note: Mapped[str] = mapped_column(Text, nullable=False)
    note_type: Mapped[str] = mapped_column(
        String(30), nullable=False, default="GENERAL", server_default="GENERAL"
    )

    vendor: Mapped[Vendor] = relationship(back_populates="notes")


class VendorAttributeValue(AttributeValueBase):
    """Store one configurable attribute value for a vendor."""

    __tablename__ = "vendor_attribute_values"
    __table_args__ = (
        UniqueConstraint(
            "vendor_id",
            "attribute_definition_id",
            name="UQ_vendor_attribute_values_owner_attribute",
        ),
        Index("IX_vendor_attribute_values_firm_text", "firm_id", "value_text"),
        Index("IX_vendor_attribute_values_firm_number", "firm_id", "value_number"),
        Index("IX_vendor_attribute_values_firm_date", "firm_id", "value_date"),
    )

    ENTITY_TYPE: ClassVar[AttributeEntityType] = AttributeEntityType.VENDOR
    OWNER_COLUMN: ClassVar[str] = "vendor_id"

    vendor_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("vendors.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
